core/templates/dev/head/domain/collection/CollectionObjectFactory.js

```diff
@@ -33,9 +33,10 @@ oppia.factory('CollectionObjectFactory', [
       // This map acts as a fast way of looking up a collection node for a given
       // exploration ID.
       this._explorationIdToNodeIndexMap = {};
-      for (var i = 0; i < collectionBackendObject.nodes.length; i++) {
+      var collectionContent = collectionBackendObject.collection_content;
```

Comment:
Again, prefer not to represent ``collection_contents`` in the actual structure of the collection (frontend or backend), just as part of the model. I think this will simplify your PR a lot.